File: python/08_018.py
class Node:
   def __init__(self, position):
      self.position = position
      self.left = None
      self.right = None
      self.height = 1

class TextEditor:
   def __init__(self):
      self.text = ""
      self.root = None
   
   def insert(self, position, text):
      if position < 0 or position > len(self.text):
         raise ValueError("Invalid position")
      
      self.text = self.text[:position] + text + self.text[position:]
      
      new_line_positions = [position + i for i, char in 
                          enumerate(text) if char == '\n']
      text_length = len(text)
      
      self._update_positions(self.root, position, text_length)
      
      for line_pos in new_line_positions:
         self.root = self._insert_node(self.root, line_pos)
   
   def delete(self, position, length):
      if position < 0 or position + length > len(self.text):
         raise ValueError("Invalid position or length")
      
      deleted_text = self.text[position:position + length]
      deleted_lines = [position + i for i, char in 
                     enumerate(deleted_text) if char == '\n']
      
      self.text = self.text[:position] + self.text[position + length:]
      
      for line_pos in deleted_lines:
         self.root = self._delete_node(self.root, line_pos)
      
      self._update_positions(self.root, position, -length)
   
   def getLineNumber(self, position):
      if position < 0 or position > len(self.text):
         raise ValueError("Invalid position")
      
      return self._count_smaller(self.root, position)
   
   def _get_height(self, node):
      if not node:
         return 0
      return node.height
   
   def _get_balance(self, node):
      if not node:
         return 0
      return self._get_height(node.left) - self._get_height(node.right)
   
   def _right_rotate(self, y):
      x = y.left
      T2 = x.right
      
      x.right = y
      y.left = T2
      
      y.height = max(self._get_height(y.left), 
                   self._get_height(y.right)) + 1
      x.height = max(self._get_height(x.left), 
                   self._get_height(x.right)) + 1
      
      return x
   
   def _left_rotate(self, x):
      y = x.right
      T2 = y.left
      
      y.left = x
      x.right = T2
      
      x.height = max(self._get_height(x.left), 
                   self._get_height(x.right)) + 1
      y.height = max(self._get_height(y.left), 
                   self._get_height(y.right)) + 1
      
      return y
   
   def _insert_node(self, node, position):
      if not node:
         return Node(position)
      
      if position < node.position:
         node.left = self._insert_node(node.left, position)
      elif position > node.position:
         node.right = self._insert_node(node.right, position)
      else:
         return node
      
      node.height = max(self._get_height(node.left), 
                      self._get_height(node.right)) + 1
      
      balance = self._get_balance(node)
      
      if balance > 1 and position < node.left.position:
         return self._right_rotate(node)
      
      if balance < -1 and position > node.right.position:
         return self._left_rotate(node)
      
      if balance > 1 and position > node.left.position:
         node.left = self._left_rotate(node.left)
         return self._right_rotate(node)
      
      if balance < -1 and position < node.right.position:
         node.right = self._right_rotate(node.right)
         return self._left_rotate(node)
      
      return node
   
   def _delete_node(self, node, position):
      if not node:
         return node
      
      if position < node.position:
         node.left = self._delete_node(node.left, position)
      elif position > node.position:
         node.right = self._delete_node(node.right, position)
      else:
         if not node.left:
            return node.right
         elif not node.right:
            return node.left
         
         successor = self._get_min_value_node(node.right)
         node.position = successor.position
         node.right = self._delete_node(node.right, 
                                      successor.position)
      
      if not node:
         return node
      
      node.height = max(self._get_height(node.left), 
                      self._get_height(node.right)) + 1
      
      balance = self._get_balance(node)
      
      if balance > 1 and self._get_balance(node.left) >= 0:
         return self._right_rotate(node)
      
      if balance > 1 and self._get_balance(node.left) < 0:
         node.left = self._left_rotate(node.left)
         return self._right_rotate(node)
      
      if balance < -1 and self._get_balance(node.right) <= 0:
         return self._left_rotate(node)
      
      if balance < -1 and self._get_balance(node.right) > 0:
         node.right = self._right_rotate(node.right)
         return self._left_rotate(node)
      
      return node
   
   def _get_min_value_node(self, node):
      current = node
      while current.left:
         current = current.left
      return current
   
   def _update_positions(self, node, start_pos, shift):
      if not node:
         return
      
      if node.position >= start_pos:
         node.position += shift
      
      self._update_positions(node.right, start_pos, shift)
      
      if node.position >= start_pos:
         self._update_positions(node.left, start_pos, shift)
   
   def _count_smaller(self, node, position):
      if not node:
         return 0
      
      if position < node.position:
         return self._count_smaller(node.left, position)
      else:
         return (1 + self._count_smaller(node.left, position) + 
                self._count_smaller(node.right, position))

File: cpp/08_018.cpp
#include <string>
#include <vector>
#include <stdexcept>
#include <algorithm>

class Node {
public:
    int position;  // Character position of the line break
    Node* left;
    Node* right;
    int height;    // For AVL tree balance
    
    Node(int pos) : position(pos), left(nullptr), 
                   right(nullptr), height(1) {}
};

class TextEditor {
private:
    std::string text;
    Node* root;  // Root of the AVL tree
    
    // AVL Tree operations
    int getHeight(Node* node) {
        if (node == nullptr) {
            return 0;
        }
        return node->height;
    }
    
    int getBalance(Node* node) {
        if (node == nullptr) {
            return 0;
        }
        return getHeight(node->left) - getHeight(node->right);
    }
    
    Node* rightRotate(Node* y) {
        Node* x = y->left;
        Node* T2 = x->right;
        
        // Perform rotation
        x->right = y;
        y->left = T2;
        
        // Update heights
        y->height = std::max(getHeight(y->left), 
                           getHeight(y->right)) + 1;
        x->height = std::max(getHeight(x->left), 
                           getHeight(x->right)) + 1;
        
        return x;
    }
    
    Node* leftRotate(Node* x) {
        Node* y = x->right;
        Node* T2 = y->left;
        
        // Perform rotation
        y->left = x;
        x->right = T2;
        
        // Update heights
        x->height = std::max(getHeight(x->left), 
                          getHeight(x->right)) + 1;
        y->height = std::max(getHeight(y->left), 
                          getHeight(y->right)) + 1;
        
        return y;
    }
    
    Node* insertNode(Node* node, int position) {
        // Standard BST insert
        if (node == nullptr) {
            return new Node(position);
        }
        
        if (position < node->position) {
            node->left = insertNode(node->left, position);
        } else if (position > node->position) {
            node->right = insertNode(node->right, position);
        } else {
            // Duplicate positions not allowed
            return node;
        }
        
        // Update height of current node
        node->height = std::max(getHeight(node->left), 
                              getHeight(node->right)) + 1;
        
        // Get the balance factor
        int balance = getBalance(node);
        
        // Left Left Case
        if (balance > 1 && 
            position < node->left->position) {
            return rightRotate(node);
        }
        
        // Right Right Case
        if (balance < -1 && 
            position > node->right->position) {
            return leftRotate(node);
        }
        
        // Left Right Case
        if (balance > 1 && 
            position > node->left->position) {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }
        
        // Right Left Case
        if (balance < -1 && 
            position < node->right->position) {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }
        
        return node;
    }
    
    Node* getMinValueNode(Node* node) {
        Node* current = node;
        while (current->left != nullptr) {
            current = current->left;
        }
        return current;
    }
    
    Node* deleteNode(Node* node, int position) {
        if (node == nullptr) {
            return node;
        }
        
        if (position < node->position) {
            node->left = deleteNode(node->left, position);
        } else if (position > node->position) {
            node->right = deleteNode(node->right, position);
        } else {
            // Node to be deleted found
            
            // Node with only one child or no child
            if (node->left == nullptr) {
                Node* temp = node->right;
                delete node;
                return temp;
            } else if (node->right == nullptr) {
                Node* temp = node->left;
                delete node;
                return temp;
            }
            
            // Node with two children
            // Get inorder successor (smallest in right subtree)
            Node* successor = getMinValueNode(node->right);
            node->position = successor->position;
            node->right = deleteNode(node->right, 
                                   successor->position);
        }
        
        // If tree had only one node
        if (node == nullptr) {
            return node;
        }
        
        // Update height
        node->height = std::max(getHeight(node->left), 
                              getHeight(node->right)) + 1;
        
        // Get balance factor
        int balance = getBalance(node);
        
        // Rebalance if needed
        // Left Left Case
        if (balance > 1 && getBalance(node->left) >= 0) {
            return rightRotate(node);
        }
        
        // Left Right Case
        if (balance > 1 && getBalance(node->left) < 0) {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }
        
        // Right Right Case
        if (balance < -1 && getBalance(node->right) <= 0) {
            return leftRotate(node);
        }
        
        // Right Left Case
        if (balance < -1 && getBalance(node->right) > 0) {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }
        
        return node;
    }
    
    void updatePositions(Node* node, int startPos, int shift) {
        if (node == nullptr) {
            return;
        }
        
        if (node->position >= startPos) {
            node->position += shift;
        }
        
        // Recursively update nodes in the right subtree
        updatePositions(node->right, startPos, shift);
        
        // Only update left subtree if current node is shifted
        if (node->position >= startPos) {
            updatePositions(node->left, startPos, shift);
        }
    }
    
    int countSmaller(Node* node, int position) {
        if (node == nullptr) {
            return 0;
        }
        
        if (position < node->position) {
            return countSmaller(node->left, position);
        } else {
            return 1 + countSmaller(node->left, position) + 
                   countSmaller(node->right, position);
        }
    }
    
    // Helper function to clean up memory
    void deleteTree(Node* node) {
        if (node == nullptr) {
            return;
        }
        
        deleteTree(node->left);
        deleteTree(node->right);
        delete node;
    }

public:
    TextEditor() : root(nullptr) {}
    
    ~TextEditor() {
        deleteTree(root);
    }
    
    void insert(int position, const std::string& str) {
        if (position < 0 || position > text.length()) {
            throw std::invalid_argument("Invalid position");
        }
        
        // Update the text
        text.insert(position, str);
        
        // Find new line breaks in the inserted text
        std::vector<int> newLinePositions;
        for (size_t i = 0; i < str.length(); i++) {
            if (str[i] == '\n') {
                newLinePositions.push_back(position + i);
            }
        }
        
        // Update existing line break positions after insertion
        updatePositions(root, position, str.length());
        
        // Insert new line breaks
        for (int linePos : newLinePositions) {
            root = insertNode(root, linePos);
        }
    }
    
    void remove(int position, int length) {
        if (position < 0 || 
            position + length > static_cast<int>(text.length())) {
            throw std::invalid_argument(
                "Invalid position or length");
        }
        
        // Count line breaks in the deleted text
        std::string deletedText = text.substr(position, length);
        std::vector<int> deletedLines;
        for (size_t i = 0; i < deletedText.length(); i++) {
            if (deletedText[i] == '\n') {
                deletedLines.push_back(position + i);
            }
        }
        
        // Update the text
        text.erase(position, length);
        
        // Delete line breaks in the deleted range
        for (int linePos : deletedLines) {
            root = deleteNode(root, linePos);
        }
        
        // Update line break positions after the deletion
        updatePositions(root, position, -length);
    }
    
    int getLineNumber(int position) {
        if (position < 0 || 
            position > static_cast<int>(text.length())) {
            throw std::invalid_argument("Invalid position");
        }
        
        // Count line breaks before the position
        return countSmaller(root, position);
    }
};